src/meili/traffic_segment_matcher.cc

```diff
@@ -261,6 +269,12 @@ std::vector<traffic_segment_t> TrafficSegmentMatcher::form_segments(const std::l
       //this is what we know so far
       //NOTE: in both cases we take the left most value for the shape index in an effort to be conservative
       traffic_segments.emplace_back(traffic_segment_t{segment->segment_id_, start_time, left->original_index, end_time, prev->original_index, length});
+
+      //if the right side of this was the end of this edge then at least we need to start from the next edge
+      if(segment->end_percent_ == 1.f) {
```

Comment:
finally another issue with loops. we use markers along the path to determine the interpolation point for the begin and end of a segment. if you do a loop that leaves an ots to non ots edges and then gets back on the same ots, previously the beginning marker didnt move, because the edge ids matched. the trick is that there was a matching edge id further along in the path. this meant that we'd report segments longer than they are by keeping the left most marker at the first occurrence of the given ots. this here shifts the markers past the edge iff the segment ends at the end of this edge. that lets us know we dont need to look at this particular edge in the path again and so we can move the markers past it, ie to the next edge.